File: tests/core_abi_smoke.zig
const std = @import("std");

const c = @cImport({
    @cInclude("melearner_core.h");
});

const max_poll_attempts = 2_000;

comptime {
    if (c.ML_ABI_VERSION != 2) @compileError("unexpected core ABI version");
    if (@sizeOf(c.ml_config_v2) != 16 + @sizeOf(usize) * 2) {
        @compileError("ml_config_v2 layout drift");
    }
    if (@offsetOf(c.ml_config_v2, "state_dir") != 16) {
        @compileError("ml_config_v2 state directory offset drift");
    }
    if (@sizeOf(c.ml_library_course_page_request_v1) != 32) {
        @compileError("course page request layout drift");
    }
    if (@sizeOf(c.ml_library_scan_request_v1) != 16 + @sizeOf(usize) * 2) {
        @compileError("scan request layout drift");
    }
    if (@sizeOf(c.ml_core_limits_v1) != 16) @compileError("ml_core_limits_v1 layout drift");
    if (@offsetOf(c.ml_event_v1, "sequence") != 8) @compileError("ml_event_v1 sequence offset drift");
    if (@offsetOf(c.ml_event_v1, "payload") != 40) @compileError("ml_event_v1 payload offset drift");
    if (@offsetOf(c.ml_event_v1, "payload_len") != 40 + @sizeOf(usize)) {
        @compileError("ml_event_v1 payload length offset drift");
    }
}

pub fn main(init: std.process.Init) !void {
    const io = init.io;
    const state_dir = try temporaryStateDirectory(init);
    defer init.gpa.free(state_dir);
    defer std.Io.Dir.cwd().deleteTree(io, state_dir) catch {};

    var config = c.ml_config_v2{
        .struct_size = @sizeOf(c.ml_config_v2),
        .abi_version = c.ML_ABI_VERSION,
        .event_queue_capacity = 4,
        .max_event_payload_bytes = 4096,
        .state_dir = state_dir.ptr,
        .state_dir_len = state_dir.len,
    };
    var core: ?*c.ml_core_t = null;
    try expectEqual(@as(c.ml_status_t, c.ML_STATUS_OK), c.ml_core_create(&config, &core));
    try expect(core != null);
    defer c.ml_core_destroy(core);
    try expectEqual(@as(u32, c.ML_ABI_VERSION), c.ml_abi_version());
    try expectEqual(@as(c.ml_status_t, c.ML_STATUS_OK), c.ml_core_set_waker(core, null, null));

    var limits = c.ml_core_limits_v1{
        .struct_size = @sizeOf(c.ml_core_limits_v1),
        .abi_version = c.ML_ABI_VERSION,
        .event_queue_capacity = 0,
        .max_event_payload_bytes = 0,
    };
    try expectEqual(@as(c.ml_status_t, c.ML_STATUS_OK), c.ml_core_get_limits_v1(core, &limits));
    try expectEqual(@as(u32, 4), limits.event_queue_capacity);
    try expectEqual(@as(u32, 4096), limits.max_event_payload_bytes);

    var ready = try pollEvent(io, core);
    try expect(ready.sequence > 0);
    try expectEqual(@as(c.ml_event_kind_t, c.ML_EVENT_CORE_READY), ready.kind);
    try expectEqual(@as(c.ml_status_t, c.ML_STATUS_OK), ready.status);
    try expectEqual(@as(u32, 1), ready.payload_schema_version);
    try expect(ready.payload != null);
    var revision = try std.fmt.parseInt(u64, ready.payload[0..ready.payload_len], 10);
    try expect(revision != 0);
    const initial_revision = revision;
    c.ml_core_release_event(core, &ready);

    var scan_request = c.ml_library_scan_request_v1{
        .struct_size = @sizeOf(c.ml_library_scan_request_v1),
        .abi_version = c.ML_ABI_VERSION,
        .expected_revision = revision,
        .root_path = state_dir.ptr,
        .root_path_len = state_dir.len,
    };
    var scan_request_id: u64 = 0;
    try expectEqual(
        @as(c.ml_status_t, c.ML_STATUS_OK),
        c.ml_library_scan_v1(core, &scan_request, &scan_request_id),
    );
    try expect(scan_request_id != 0);

    var scanned = try pollEvent(io, core);
    try expectEqual(scan_request_id, scanned.request_id);
    try expectEqual(@as(c.ml_event_kind_t, c.ML_EVENT_LIBRARY_SCAN), scanned.kind);
    try expectEqual(@as(c.ml_status_t, c.ML_STATUS_OK), scanned.status);
    try expectEqual(@as(u32, 1), scanned.payload_schema_version);
    try expect(scanned.payload != null);
    const scan_payload = scanned.payload[0..scanned.payload_len];
    const scan_prefix = "{\"revision\":";
    const scan_suffix = ",\"courseCount\":0,\"warnings\":[]}";
    try expect(std.mem.startsWith(u8, scan_payload, scan_prefix));
    try expect(std.mem.endsWith(u8, scan_payload, scan_suffix));
    revision = try std.fmt.parseInt(
        u64,
        scan_payload[scan_prefix.len .. scan_payload.len - scan_suffix.len],
        10,
    );
    try expect(revision > initial_revision);
    c.ml_core_release_event(core, &scanned);

    var request = c.ml_library_course_page_request_v1{
        .struct_size = @sizeOf(c.ml_library_course_page_request_v1),
        .abi_version = c.ML_ABI_VERSION,
        .expected_revision = revision,
        .offset = 0,
        .limit = 20,
        .reserved = 0,
    };
    var request_id: u64 = 0;
    try expectEqual(
        @as(c.ml_status_t, c.ML_STATUS_OK),
        c.ml_library_course_page_v1(core, &request, &request_id),
    );
    try expect(request_id != 0);

    var completed = try pollEvent(io, core);
    try expectEqual(request_id, completed.request_id);
    try expectEqual(@as(c.ml_event_kind_t, c.ML_EVENT_LIBRARY_COURSE_PAGE), completed.kind);
    try expectEqual(@as(c.ml_status_t, c.ML_STATUS_OK), completed.status);
    try expectEqual(@as(u32, 1), completed.payload_schema_version);
    const expected_page = try std.fmt.allocPrint(
        init.gpa,
        "{{\"revision\":{d},\"offset\":0,\"total\":0,\"rows\":[]}}",
        .{revision},
    );
    defer init.gpa.free(expected_page);
    try expectPayload(&completed, expected_page);
    c.ml_core_release_event(core, &completed);

    var empty = event();
    try expectEqual(@as(c.ml_status_t, c.ML_STATUS_EMPTY), c.ml_core_poll_event(core, &empty));
}

fn temporaryStateDirectory(init: std.process.Init) ![]u8 {
    var random_bytes: [16]u8 = undefined;
    std.Io.random(init.io, &random_bytes);
    const suffix = std.fmt.bytesToHex(random_bytes, .lower);
    const temporary_root = init.environ_map.get("TMPDIR") orelse "/tmp";
    const state_dir = try std.fmt.allocPrint(
        init.gpa,
        "{s}/melearner-core-abi-{s}",
        .{ temporary_root, suffix },
    );
    errdefer init.gpa.free(state_dir);
    try std.Io.Dir.cwd().createDir(init.io, state_dir, .default_dir);
    return state_dir;
}

fn pollEvent(io: std.Io, core: ?*c.ml_core_t) !c.ml_event_v1 {
    for (0..max_poll_attempts) |_| {
        var next = event();
        switch (c.ml_core_poll_event(core, &next)) {
            c.ML_STATUS_OK => return next,
            c.ML_STATUS_EMPTY => try std.Io.sleep(
                io,
                std.Io.Duration.fromMilliseconds(1),
                .awake,
            ),
            else => return error.CoreAbiSmokeFailed,
        }
    }
    return error.CoreAbiSmokeTimedOut;
}

fn event() c.ml_event_v1 {
    return .{
        .struct_size = @sizeOf(c.ml_event_v1),
        .abi_version = c.ML_ABI_VERSION,
        .sequence = 0,
        .request_id = 0,
        .kind = 0,
        .status = 0,
        .payload_schema_version = 0,
        .reserved = 0,
        .payload = null,
        .payload_len = 0,
    };
}

fn expectPayload(event_value: *const c.ml_event_v1, expected: []const u8) !void {
    try expect(event_value.payload != null);
    try expectEqual(expected.len, event_value.payload_len);
    try expect(std.mem.eql(u8, expected, event_value.payload[0..event_value.payload_len]));
}

fn expect(condition: bool) !void {
    if (!condition) return error.CoreAbiSmokeFailed;
}

fn expectEqual(expected: anytype, actual: @TypeOf(expected)) !void {
    if (actual != expected) return error.CoreAbiSmokeFailed;
}

File: tests/core_abi_smoke.c
#include "melearner_core.h"

#include <errno.h>
#include <stddef.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <threads.h>

#if !defined(__STDC_VERSION__) || __STDC_VERSION__ < 202311L
#error "melearner C ABI smoke test requires ISO C23"
#endif

constexpr uint32_t expected_abi = 2;
constexpr size_t max_poll_attempts = 2'000'000;

static_assert(ML_ABI_VERSION == expected_abi);
static_assert(ML_STATUS_OK == 0);
static_assert(ML_STATUS_TOO_LATE == 11);
static_assert(ML_EVENT_CORE_READY == 1);
static_assert(ML_EVENT_LIBRARY_STATS == 15);
static_assert(ML_EVENT_LIBRARY_STATE == 16);
static_assert(ML_EVENT_SETTINGS == 17);
static_assert(ML_EVENT_APPEARANCE_UPDATED == 18);
static_assert(ML_EVENT_DOCUMENT_OPENED == 19);
static_assert(ML_EVENT_DOCUMENT_PAGE == 20);
static_assert(ML_EVENT_DOCUMENT_EXTERNAL_OPEN_READY == 21);
static_assert(ML_APPEARANCE_LIGHT == 1);
static_assert(ML_APPEARANCE_DARK == 2);
static_assert(ML_APPEARANCE_COZY == 3);
static_assert(ML_SCAN_PHASE_DISCOVERING == 1);
static_assert(ML_SCAN_PHASE_WRITING_MARKERS == 5);
static_assert(sizeof(ml_config_v2) == 16 + sizeof(size_t) * 2);
static_assert(offsetof(ml_config_v2, state_dir) == 16);
static_assert(sizeof(ml_library_stats_request_v1) == 24);
static_assert(offsetof(ml_library_stats_request_v1, expected_revision) == 8);
static_assert(sizeof(ml_library_state_request_v1) == 24);
static_assert(offsetof(ml_library_state_request_v1, expected_revision) == 8);
static_assert(sizeof(ml_settings_get_request_v1) == 16);
static_assert(offsetof(ml_settings_get_request_v1, reserved) == 8);
static_assert(sizeof(ml_settings_put_appearance_request_v1) == 24);
static_assert(offsetof(ml_settings_put_appearance_request_v1, expected_revision) == 8);
static_assert(offsetof(ml_settings_put_appearance_request_v1, appearance) == 16);
static_assert(offsetof(ml_settings_put_appearance_request_v1, reserved) == 20);
static_assert(sizeof(ml_document_open_request_v1) == 24 + sizeof(size_t) * 2);
static_assert(offsetof(ml_document_open_request_v1, expected_revision) == 8);
static_assert(offsetof(ml_document_open_request_v1, reserved) == 16);
static_assert(offsetof(ml_document_open_request_v1, lesson_id) == 24);
static_assert(sizeof(ml_document_page_request_v1) == 32 + sizeof(size_t) * 2);
static_assert(offsetof(ml_document_page_request_v1, expected_revision) == 8);
static_assert(offsetof(ml_document_page_request_v1, offset) == 16);
static_assert(offsetof(ml_document_page_request_v1, limit) == 24);
static_assert(offsetof(ml_document_page_request_v1, reserved) == 28);
static_assert(offsetof(ml_document_page_request_v1, document_id) == 32);
static_assert(sizeof(ml_document_external_open_request_v1) == 24 + sizeof(size_t) * 2);
static_assert(offsetof(ml_document_external_open_request_v1, lesson_id) == 24);
static_assert(sizeof(ml_library_scan_progress_snapshot_v1) == 48);
static_assert(offsetof(ml_library_scan_progress_snapshot_v1, request_id) == 8);
static_assert(offsetof(ml_library_scan_progress_snapshot_v1, processed) == 16);
static_assert(offsetof(ml_library_scan_progress_snapshot_v1, phase) == 40);
static_assert(offsetof(ml_library_scan_progress_snapshot_v1, total_known) == 44);
static_assert(offsetof(ml_library_scan_progress_snapshot_v1, reserved) == 46);
static_assert(offsetof(ml_event_v1, sequence) == 8);
static_assert(offsetof(ml_event_v1, payload) == 40);
static_assert(offsetof(ml_event_v1, payload_len) == 40 + sizeof(void *));

#define CHECK(expression)                                                                          \
  do {                                                                                             \
    if (!(expression)) {                                                                           \
      fprintf(stderr, "check failed at %s:%d: %s\n", __FILE__, __LINE__, #expression);             \
      return EXIT_FAILURE;                                                                         \
    }                                                                                              \
  } while (0)

static ml_status_t poll_event(ml_core_t *core, ml_event_v1 *event) {
  *event = (ml_event_v1){
      .struct_size = sizeof(*event),
      .abi_version = ML_ABI_VERSION,
  };
  for (size_t attempt = 0; attempt < max_poll_attempts; ++attempt) {
    const ml_status_t status = ml_core_poll_event(core, event);
    if (status != ML_STATUS_EMPTY) {
      return status;
    }
    thrd_yield();
  }
  return ML_STATUS_EMPTY;
}

static bool copy_payload(const ml_event_v1 *event, char *buffer, size_t capacity) {
  if (event->payload == nullptr || event->payload_len == 0 || event->payload_len >= capacity) {
    return false;
  }
  memcpy(buffer, event->payload, event->payload_len);
  buffer[event->payload_len] = '\0';
  return true;
}

static bool parse_revision(const ml_event_v1 *event, uint64_t *revision) {
  char text[32];
  if (!copy_payload(event, text, sizeof(text))) {
    return false;
  }
  char *end = nullptr;
  errno = 0;
  const unsigned long long value = strtoull(text, &end, 10);
  if (errno != 0 || end != text + event->payload_len || value == 0) {
    return false;
  }
  *revision = (uint64_t)value;
  return true;
}

int main(int argc, char **argv) {
  CHECK(argc == 2);
  const char *state_dir = argv[1];
  const size_t state_dir_len = strlen(state_dir);
  CHECK(state_dir_len != 0);

  const ml_config_v2 config = {
      .struct_size = sizeof(config),
      .abi_version = ML_ABI_VERSION,
      .event_queue_capacity = 4,
      .max_event_payload_bytes = 4096,
      .state_dir = (const uint8_t *)state_dir,
      .state_dir_len = state_dir_len,
  };
  ml_config_v2 incompatible_config = config;
  incompatible_config.abi_version = ML_ABI_VERSION + 1;
  ml_core_t *incompatible_core = nullptr;
  CHECK(ml_core_create(&incompatible_config, &incompatible_core) == ML_STATUS_ABI_MISMATCH);
  CHECK(incompatible_core == nullptr);

  ml_core_t *core = nullptr;
  CHECK(ml_abi_version() == expected_abi);
  CHECK(ml_core_create(&config, &core) == ML_STATUS_OK);
  CHECK(core != nullptr);
  CHECK(ml_core_set_waker(core, nullptr, nullptr) == ML_STATUS_OK);

  ml_core_limits_v1 limits = {
      .struct_size = sizeof(limits),
      .abi_version = ML_ABI_VERSION,
  };
  CHECK(ml_core_get_limits_v1(core, &limits) == ML_STATUS_OK);
  CHECK(limits.event_queue_capacity == config.event_queue_capacity);
  CHECK(limits.max_event_payload_bytes == config.max_event_payload_bytes);

  ml_event_v1 event;
  CHECK(poll_event(core, &event) == ML_STATUS_OK);
  CHECK(event.sequence != 0);
  CHECK(event.kind == ML_EVENT_CORE_READY);
  CHECK(event.status == ML_STATUS_OK);
  CHECK(event.payload_schema_version == 1);
  uint64_t revision = 0;
  CHECK(parse_revision(&event, &revision));
  ml_core_release_event(core, &event);

  const ml_library_state_request_v1 state_request = {
      .struct_size = sizeof(state_request),
      .abi_version = ML_ABI_VERSION,
      .expected_revision = revision,
      .reserved = 0,
  };
  uint64_t request_id = 0;
  CHECK(ml_library_state_v1(core, &state_request, &request_id) == ML_STATUS_OK);
  CHECK(request_id != 0);
  CHECK(poll_event(core, &event) == ML_STATUS_OK);
  CHECK(event.request_id == request_id);
  CHECK(event.kind == ML_EVENT_LIBRARY_STATE);
  CHECK(event.status == ML_STATUS_OK);
  char state_payload[128];
  CHECK(copy_payload(&event, state_payload, sizeof(state_payload)));
  CHECK(strstr(state_payload, "\"rootPath\":null") != nullptr);
  ml_core_release_event(core, &event);

  constexpr char missing_lesson[] = "missing-document";
  const ml_document_open_request_v1 document_request = {
      .struct_size = sizeof(document_request),
      .abi_version = ML_ABI_VERSION,
      .expected_revision = revision,
      .reserved = 0,
      .lesson_id = (const uint8_t *)missing_lesson,
      .lesson_id_len = sizeof(missing_lesson) - 1,
  };
  request_id = 0;
  CHECK(ml_document_open_v1(core, &document_request, &request_id) == ML_STATUS_OK);
  CHECK(poll_event(core, &event) == ML_STATUS_OK);
  CHECK(event.request_id == request_id);
  CHECK(event.kind == ML_EVENT_DOCUMENT_OPENED);
  CHECK(event.status == ML_STATUS_NOT_FOUND);
  char document_payload[128];
  CHECK(copy_payload(&event, document_payload, sizeof(document_payload)));
  CHECK(strcmp(document_payload, "{\"error\":\"lessonNotFound\"}") == 0);
  ml_core_release_event(core, &event);

  const ml_settings_get_request_v1 settings_request = {
      .struct_size = sizeof(settings_request),
      .abi_version = ML_ABI_VERSION,
      .reserved = 0,
  };
  request_id = 0;
  CHECK(ml_settings_get_v1(core, &settings_request, &request_id) == ML_STATUS_OK);
  CHECK(poll_event(core, &event) == ML_STATUS_OK);
  CHECK(event.request_id == request_id);
  CHECK(event.kind == ML_EVENT_SETTINGS);
  CHECK(event.status == ML_STATUS_OK);
  char settings_payload[128];
  CHECK(copy_payload(&event, settings_payload, sizeof(settings_payload)));
  CHECK(strcmp(settings_payload, "{\"revision\":1,\"appearance\":\"light\"}") == 0);
  ml_core_release_event(core, &event);

  const ml_settings_put_appearance_request_v1 appearance_request = {
      .struct_size = sizeof(appearance_request),
      .abi_version = ML_ABI_VERSION,
      .expected_revision = 1,
      .appearance = ML_APPEARANCE_COZY,
      .reserved = 0,
  };
  request_id = 0;
  CHECK(ml_settings_put_appearance_v1(core, &appearance_request, &request_id) == ML_STATUS_OK);
  CHECK(poll_event(core, &event) == ML_STATUS_OK);
  CHECK(event.request_id == request_id);
  CHECK(event.kind == ML_EVENT_APPEARANCE_UPDATED);
  CHECK(event.status == ML_STATUS_OK);
  CHECK(copy_payload(&event, settings_payload, sizeof(settings_payload)));
  CHECK(strcmp(settings_payload, "{\"revision\":2,\"appearance\":\"cozy\"}") == 0);
  ml_core_release_event(core, &event);

  const ml_library_stats_request_v1 stats_request = {
      .struct_size = sizeof(stats_request),
      .abi_version = ML_ABI_VERSION,
      .expected_revision = revision,
      .reserved = 0,
  };
  request_id = 0;
  CHECK(ml_library_stats_v1(core, &stats_request, &request_id) == ML_STATUS_OK);
  CHECK(request_id != 0);
  CHECK(poll_event(core, &event) == ML_STATUS_OK);
  CHECK(event.request_id == request_id);
  CHECK(event.kind == ML_EVENT_LIBRARY_STATS);
  CHECK(event.status == ML_STATUS_OK);
  CHECK(event.payload_schema_version == 1);

  char stats_payload[1024];
  CHECK(copy_payload(&event, stats_payload, sizeof(stats_payload)));
  CHECK(strstr(stats_payload, "\"totalCourses\":0") != nullptr);
  CHECK(strstr(stats_payload, "\"mediaTypes\":[]") != nullptr);
  CHECK(strstr(stats_payload, "\"topCourses\":[]") != nullptr);
  ml_core_release_event(core, &event);

  ml_core_destroy(core);
  ml_core_destroy(nullptr);
  return EXIT_SUCCESS;
}
